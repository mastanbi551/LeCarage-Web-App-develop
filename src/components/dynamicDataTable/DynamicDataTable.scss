@import "../../styles/variables.scss";

.pagination {
  display: flex;
  justify-content: space-between;
  list-style: none;
  cursor: pointer;
  margin: 5px;
  align-self: flex-end;
}

.pagination a {
  text-decoration: none;
  padding: 5px;
  color: $veryDarkGray;
}

.pagination__link {
  font-weight: bold;
}

.pagination__link--active a {
  margin: 5px;
  padding: 7px;
  color: $primaryWhite !important;
  background: $veryDarkGray;
  border-radius: 28%;
}

.pagination__link--disabled a {
  color: $darkgrey;
}

.card {
  background-color: $White;
  padding: 5px;
  border: 0px;
}

.custom-pagination {
  display: flex;
  justify-content: space-between;
  list-style: none;
  cursor: pointer;
  margin: 5px;
  align-self: flex-end;
  padding: 0px 10px;

  .rows-per-page {
    display: block;

    .label {
      vertical-align: middle;
      font-family: $montserratFont;
      font-style: normal;
      font-weight: 700;
      font-size: 12px;
      line-height: 29px;
      letter-spacing: 0.3px;
      color: $primaryBlack;
      display: inline-block;
    }

    select {
      @include sidebar_links;
      border: none;
      font-family: $latoFont;
      letter-spacing: 0.1px;
      color: $veryDarkGray;
      display: inline-block;
    }

    .dropdown {
      display: inline-block;
    }
    select {
      width: 50px;
      padding: 5px;
      font-size: 14px;
      line-height: 1;
      border: 0;
      border-radius: 5px;
      height: 34px;
      background: url("data:image/svg+xml,<svg fill='currentColor' color='grey' height='10px' width='10px' viewBox='0 0 16 16'  xmlns='http://www.w3.org/2000/svg'><path  d='M7.247 11.14 2.451 5.658C1.885 5.013 2.345 4 3.204 4h9.592a1 1 0 0 1 .753 1.659l-4.796 5.48a1 1 0 0 1-1.506 0z'/></svg>")
        no-repeat right;
      appearance: none;
      background-position-x: 25px;
      background-size: 14px;
    }
  }
  .showing-arrows {
    display: block !important;

    .label {
      vertical-align: middle;
      font-family: $montserratFont;
      font-style: normal;
      font-weight: 700;
      font-size: 12px;
      line-height: 29px;
      letter-spacing: 0.3px;
      color: $primaryBlack;
    }

    label,
    .buttons {
      display: inline-block;
      padding-right: 10px;
    }

    .buttons button {
      border: none !important;
      background: none;
      margin: 5px;
      color: $black;
    }

    button[disabled] {
      color: $disablebtn;
    }
  }
}

.dynamic-table {
  thead {
    th {
      @include tableHeading;
    }
  }
  tbody {
    td {
      @include tableBody;
      padding: 15px 0px 15px 20px;
      cursor: pointer;
    }

    .edit-delete-icons svg {
      width: 17px;
      height: 17px;
      margin-right: 7px;
    }
  }
}

.pt {
  padding-top: 10px;
  padding-left: 18px;
}
.norecord {
  padding: 5px;
  font-size: 17px;
}

.accessdisabled{
  
  color: $darkgrey;
  cursor: not-allowed;  
}