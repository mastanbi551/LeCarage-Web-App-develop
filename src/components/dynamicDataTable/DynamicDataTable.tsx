import FuzzySearch from 'fuzzy-search';
import React, { useEffect, useState } from 'react';
import { DynamicDataTableItems } from './DynamicDataTableItems';
import './DynamicDataTable.scss';
import _ from 'lodash';

interface DynamicTableProps {
  tableHeading: string;
  isSearchIconEnable: boolean;
  isFilterIconEnable: boolean;
  columns: {accessor: string, Header: string}[];
  data: [];
  searchValue: string[];
  filterValues: {mainMenu: string, subMenus: string[]}[];
  onViewHandler: (rowInfo: object) => void;
  onEditHandler: (rowInfo: object) => void;
  onDeleteHandler?: (rowInfo: object) => void;
  defaultItemsPerPage: number;
  errorMessage?: string;
  isEditIconEnable?: boolean;
  isDeleteIconEnable?: boolean;
  access?: string;
  rowsPerPageValueChange: (value: string) => void;
}

export const DynamicDataTable = ({
  tableHeading,
  isSearchIconEnable,
  isFilterIconEnable,
  columns,
  data,
  searchValue,
  filterValues,
  onViewHandler,
  onEditHandler,
  onDeleteHandler,
  defaultItemsPerPage,
  errorMessage,
  isEditIconEnable,
  isDeleteIconEnable,
  access,
  rowsPerPageValueChange
}: DynamicTableProps) => {

  const [itemOffset, setItemOffset] = useState(0);
  const [pageData, setPageData] = useState([]);
  const [customData, setCustomData] = useState([]);

  const [endOffset, setEndOffset] = useState(0);
  const [pageIndex, setPageIndex] = useState(0);
  const [itemsPerPage, setItemsPerPage] = useState(5);

  const [rowsPerPage, setRowsPerPage] = useState([]);
  const [displayColumns, setDisplayColumns] = useState<{label: string, value: string}[]>([]);
  
  const setRows = (rowsPerPageData: never[]) => {
    const columns: {label: string, value: string}[] = [];
    const chunkedArray = _.chunk(rowsPerPageData, itemsPerPage);
    let pagePerItems = 0;
    rowsPerPageData && chunkedArray && chunkedArray.map(() => {
      if(displayColumns.length === 0){
        pagePerItems = pagePerItems + parseInt(itemsPerPage.toString());
        const updatedColumn = [{
            label: pagePerItems.toString(),
            value: pagePerItems.toString(),
          }
        ];
        columns.push(updatedColumn[0]);
      }
      else {
        pagePerItems = pagePerItems + parseInt(itemsPerPage.toString());
        const updatedColumn = [
          ...displayColumns, {
            label: pagePerItems.toString(),
            value: pagePerItems.toString(),
          }
        ];
        columns.push(updatedColumn[0]);
      }
    });
    setDisplayColumns(columns);
  };

  useEffect(() => {
    if(data.length > 0){
      setCustomData(data);
    }
  }, [data]);

  useEffect(() => {
    if(customData && customData.length > 0 && displayColumns.length === 0){
      setRows(customData);
    }
  }, [customData]);

  const updateData = (staticData: never[]) => {
    setEndOffset(itemOffset + parseInt(defaultItemsPerPage?.toString()));
    setPageData(staticData?.slice(itemOffset, endOffset));
  };

  useEffect(() => {
    updateData(customData);
  }, [endOffset, itemOffset, defaultItemsPerPage]);


  const searcher = new FuzzySearch(data, searchValue, {
    caseSensitive: false,
  });

  const filter = new FuzzySearch(data, searchValue, {
    caseSensitive: true,
  });

  const setData = (data: never[]) => {
    setPageData(data);
    setItemsPerPage(defaultItemsPerPage);
    setItemOffset(0);
    updateData(data);
  };

  const onSearchChangeHandler = (searchTerm: string) => {
    const result = searcher.search(searchTerm);    

    const filteredArray = data.filter((obj) => {
      const values = Object.values(obj);
      return values.some((value: any) =>
        value !== null && value.toString().toLowerCase().includes(searchTerm.toLowerCase())
      );
    });

    setData(filteredArray);
    setCustomData(filteredArray);
    setDisplayColumns([]);
  };

  const handleCategoryChange = (value: string) => {
    if(value !== ''){
      const result = filter.search(value);
      setData(result);
      setCustomData(result);
      setDisplayColumns([]);
    }
  };

  const currentPage = pageIndex + 1;
  const canPreviousPage = pageIndex > 0;
  const canNextPage = currentPage < Math.ceil(data.length / defaultItemsPerPage);

  useEffect(() => {
    if(pageIndex === 0){
      const newOffset = (pageIndex * parseInt(defaultItemsPerPage.toString())) % data.length;
      setItemOffset(newOffset);
    }
    else {
      const newOffset = (pageIndex * parseInt(defaultItemsPerPage.toString())) % data.length;
      setItemOffset(newOffset);
    }
  }, [pageIndex, data]);

  const nextPage = () => {
    setPageIndex(pageIndex + 1);
  };

  const previousPage = () => {
    setPageIndex(pageIndex - 1);
  };

  const gotoPage = (page: number) => {
    if (page == 0) {
      setPageIndex(page);
    } else {
      setPageIndex(page);
    }
  };

  return (
    <>
      <DynamicDataTableItems
        tableHeading={tableHeading}
        isSearchIconEnable={isSearchIconEnable}
        isFilterIconEnable={isFilterIconEnable}
        columns={columns}
        data={pageData}
        tableData={customData}
        filterValues={filterValues}
        onViewHandler={onViewHandler}
        onEditHandler={onEditHandler}
        onDeleteHandler={onDeleteHandler}
        onSearchChangeHandler={onSearchChangeHandler}
        handleCategoryChange={handleCategoryChange}
        rowsPerPage={displayColumns}
        rowsPerPageValueChange={rowsPerPageValueChange}        
        itemOffset={itemOffset}
        endOffset={endOffset}       
        gotoPage={gotoPage}
        previousPage={previousPage}
        nextPage={nextPage}
        canPreviousPage={canPreviousPage}
        canNextPage={canNextPage}
        errorMessage={errorMessage}
        isEditIconEnable={isEditIconEnable}
        isDeleteIconEnable={isDeleteIconEnable}
        access={access}
        defaultItemsPerPage={defaultItemsPerPage}
      />
    </>
  );
};
