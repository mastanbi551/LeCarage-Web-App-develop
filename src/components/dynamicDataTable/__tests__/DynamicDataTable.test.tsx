import React from 'react';
import { render, screen } from '@testing-library/react';
import { DynamicDataTable } from '../DynamicDataTable';

describe('DynamicDataTable', () => {
  const columns = [
    { Header: 'Name', accessor: 'name' },
    { Header: 'Age', accessor: 'age' },
  ];
  const data = [
    { name: 'John', age: 30 },
    { name: 'Jane', age: 25 },
    { name: 'Bob', age: 40 },
  ];

  it('should render table heading', () => {
    const tableHeading = 'Test Table';
    render(
      <DynamicDataTable
        tableHeading={tableHeading}
        isSearchIconEnable={true}
        isFilterIconEnable={true}
        columns={columns}
        data={data}
        defaultItemsPerPage={5}
      />
    );
    expect(screen.getByText(tableHeading)).toBeInTheDocument();
  });

  it('should render table with data', () => {
    render(
      <DynamicDataTable
        tableHeading="Test Table"
        isSearchIconEnable={true}
        isFilterIconEnable={true}
        columns={columns}
        data={data}
        defaultItemsPerPage={5}
      />
    );
    expect(screen.getByText('John')).toBeInTheDocument();
    expect(screen.getByText('Jane')).toBeInTheDocument();
    expect(screen.getByText('Bob')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <DynamicDataTable
        tableHeading="Test Table"
        isSearchIconEnable={true}
        isFilterIconEnable={true}
        columns={columns}
        data={data}
        defaultItemsPerPage={5}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
