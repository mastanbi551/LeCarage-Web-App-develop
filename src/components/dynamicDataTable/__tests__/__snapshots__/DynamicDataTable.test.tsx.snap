// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicDataTable should match snapshot 1`] = `
<div
  class="user-types__table-header responsive"
>
  <div
    class="table-header"
  >
    <h2
      class="table-header__heading"
    >
      Test Table
    </h2>
    <div
      class="table-header__icons"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
        />
      </svg>
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="22 3 2 3 10 12.46 10 19 14 21 14 12.46 22 3"
        />
      </svg>
    </div>
  </div>
</div>
`;
